### PR TITLE
Add support for cluster authentication mode and access entries

### LINKED ISSUE #1027: Support ClusterAccessConfig on EKS Cluster
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Want to be able to setup ClusterAccessConfig to adopt the new authentication method with users into K8s to move off the config map model.

### Affected area/feature

EKS authentication/Cluster Features

### PR DESCRIPTION
**Pull request by flostadler**
### Proposed changes

AWS recently introduced a new method for granting IAM principals access to Kubernetes resources called[ Access Entries](https://docs.aws.amazon.com/eks/latest/userguide/grant-k8s-access.html#authentication-modes). These resources are now the recommended approach for controlling access to EKS clusters.

Previously, the aws-auth ConfigMap was the sole method for mapping IAM principals to Kubernetes RBAC. Now, users can choose between using the ConfigMap (access mode CONFIG_MAP), Access Entries (access mode API), or both (access mode API_AND_CONFIG_MAP).

This PR adds support for the new access modes as described in this internal doc: https://docs.google.com/document/d/1QS7h2E6lVTf8F6eVoHJOX3KowufvbhLqgRFFzTSb-SU/edit#heading=h.3b4er6mf60f5

### How to review?
Start by reviewing the schema changes ([provider/cmd/pulumi-gen-eks/main.go](https://github.com/pulumi/pulumi-eks/pull/1171/files#diff-722023496cc06082723bd16199e11061dcfbfe49e22c151096c2c524b5578d76)) and then have a look at the changes to the cluster ([nodejs/eks/cluster.ts](https://github.com/pulumi/pulumi-eks/pull/1171/files#diff-6a4020d3ae620ecc22f9d2891adeaf8761dcfb2a337f41e03156e4b815f60f13)).

After that have a look at the user facing documentation for migrating between access modes ([docs/authentication-mode-migration.md](https://github.com/pulumi/pulumi-eks/pull/1171/files#diff-fecf407cb6de248cbc2ad0be57b2423c504154fb5038d13021b7b67be617ed34)) and the accompanying test for it ([examples/tests/authentication-mode-migration/README.md](https://github.com/pulumi/pulumi-eks/pull/1171/files#diff-0598f3abbc7fd17a734e634f464755e1b9713225a00227731e42837ac832672a)).

### Related issues (optional)

Closes #1027

### CODE REVIEW

File: examples/utils/utils.go [hunk 1]
```diff
@@ -533,14 +570,55 @@ func mapClusterToKubeAccess(kubeconfigs ...interface{}) (clusterKubeAccessMap, e
 		}
 
 		// Parse the kubeconfig user auth exec args for the cluster name.
-		clusterNameIndex := len(kubeAccess.RESTConfig.ExecProvider.Args) - 1
+		var clusterNameIndex int
```
Comment (flostadler) [inline examples/utils/utils.go:573]: This previously assumed that the cluster name is always the last parameter in the list, but that's not true (with the current aws cli version it's the output type for example).

This now retrieves the correct positional argument

File: nodejs/eks/authenticationMode.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+ * @param userMappings - The IAM user mappings to be included.
+ * @returns The AWS authentication data for the aws-auth ConfigMap.
+ * @throws Error if the IAM role mappings or user mappings are invalid or cannot be serialized to YAML.
+ */
+export function createAwsAuthData(
+    instanceRoles: pulumi.Output<aws.iam.Role[]>,
+    roleMappings: pulumi.Input<pulumi.Input<RoleMapping>[]> | undefined,
+    userMappings: pulumi.Input<pulumi.Input<UserMapping>[]> | undefined,
+): pulumi.Input<{ [key: string]: pulumi.Input<string> }> {
+    const instanceRoleMappings = instanceRoles.apply((roles) =>
```
Comment (flostadler) [inline nodejs/eks/authenticationMode.ts:94]: This is the same as before, just extracted it to here to slim down the `cluster.ts` a bit and make it more manageable

File: nodejs/eks/nodegroup.ts [hunk 3]
```diff
@@ -1231,9 +1231,12 @@ ${customUserData}
             metadataOptions: args.metadataOptions,
             userData: userdata,
             tagSpecifications: args.launchTemplateTagSpecifications,
-            monitoring: {
-                enabled: args.enableDetailedMonitoring,
-            },
+            monitoring:
```
Comment (flostadler) [inline nodejs/eks/nodegroup.ts:1234]: This changed by pulling the new upstream provider version. Without this we're always getting diffs when the `monitoring` block is empty (i.e. all it's properties are undefined`

File: nodejs/eks/package.json [hunk 4]
```diff
@@ -19,7 +19,7 @@
     },
     "bugs": "https://github.com/pulumi/pulumi-eks/issues",
     "dependencies": {
-        "@pulumi/aws": "^6.0.4",
+        "@pulumi/aws": "^6.18.2",
```
Comment (flostadler) [inline nodejs/eks/package.json:22]: This is the first version supporting access entries

File: provider/cmd/pulumi-gen-eks/main.go [hunk 5]
```diff
@@ -1427,6 +1457,81 @@ func generateSchema() schema.PackageSpec {
...
+								AdditionalProperties: &schema.TypeSpec{Type: "string"},
+							},
+							Description: "The tags to apply to the AccessEntry.",
+						},
+						"type": {
+							TypeSpec: schema.TypeSpec{
+								Type: "string",
+							},
+							Description: "The type of the new access entry. Valid values are STANDARD, FARGATE_LINUX, EC2_LINUX, and EC2_WINDOWS.\n" +
+								"Defaults to STANDARD which provides the standard workflow. EC2_LINUX, EC2_WINDOWS, FARGATE_LINUX types disallow users to input a username or kubernetesGroup, and prevent associating access policies.",
```
Comment (blampe) [inline provider/cmd/pulumi-gen-eks/main.go:1506]: Also might make more sense as an enum.

File: provider/cmd/pulumi-gen-eks/main.go [hunk 6]
```diff
@@ -643,6 +643,27 @@ func generateSchema() schema.PackageSpec {
...
+						Description: "Access entries to add to the EKS cluster. They can be used to allow IAM principals to access the cluster. " +
+							"Access entries are only supported with authentication mode `API` or `API_AND_CONFIG_MAP`.\n\n" +
+							"See for more details:\nhttps://docs.aws.amazon.com/eks/latest/userguide/access-entries.html",
+					},
+					"authenticationMode": {
+						TypeSpec: schema.TypeSpec{
+							Type:  "string",
+							Plain: true,
+						},
+						Description: "The authentication mode of the cluster. Valid values are `CONFIG_MAP`, `API` or `API_AND_CONFIG_MAP`.\n\n" +
```
Comment (blampe) [inline provider/cmd/pulumi-gen-eks/main.go:664]: This might be appropriate for an enum.

**Review by blampe — COMMENTED** [review]
This looks reasonable to me but @rquitales should definitely have a look!

File: docs/authentication-mode-migration.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
+# Authentication Mode Migration
```
Comment (t0yv0) [inline docs/authentication-mode-migration.md:1]: This is great! Link to release notes please.

File: provider/cmd/pulumi-gen-eks/main.go [hunk 8]
```diff
@@ -1460,7 +1565,7 @@ func generateSchema() schema.PackageSpec {
 			}),
 			"java": rawMessage(map[string]interface{}{
 				"dependencies": map[string]string{
-					"com.pulumi:aws":        "6.5.0",
+					"com.pulumi:aws":        "6.18.2",
```
Comment (t0yv0) [inline provider/cmd/pulumi-gen-eks/main.go:1568]: https://github.com/pulumi/pulumi-eks/issues/1175 perhaps we could keep it updated more frequently. If we're getting better with keeping AWS backwards-compatible it's probably a good thing to do.

File: nodejs/eks/nodegroup.ts (hunk 3 above)
Comment (t0yv0) [reply nodejs/eks/nodegroup.ts:1234]: You can throw an issue in pulumi-aws if relevant happy to dig deep into that.

**Comment by t0yv0** [comment]
Looks good at first glance! Consider enums 🙏 

I was surprised to see there is no sdk/nodejs this is very interesting.

File: nodejs/eks/nodegroup.ts (hunk 3 above)
Comment (flostadler) [reply nodejs/eks/nodegroup.ts:1234]: That's one of the permadiffs on empty objects that we've already found some of (empty blocks are not saved to state).

I can cut a ticket to look into it holistically

File: provider/cmd/pulumi-gen-eks/main.go (hunk 5 above)
Comment (flostadler) [reply provider/cmd/pulumi-gen-eks/main.go:1506]: Sounds good!

File: docs/authentication-mode-migration.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+- `userMappings`
+- `instanceRoles`
+
+To create Access Entries for those you need to choose the right Access Entry type.
+
+#### `roleMappings` & `userMappings`
+These map to Access Entries of type `STANDARD`, which is the default.
+
+The following shows how you'd add access entries for the role and user mappings:
+```
```
Comment (corymhall) [inline docs/authentication-mode-migration.md:41]:
```suggestion
```ts
```

File: docs/authentication-mode-migration.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    ...
+});
+```
+
+#### `instanceRoles`
+These map to Access Entries of type `EC2_LINUX`.
+
+The following shows how you'd add access entries for the instance roles:
+```
```
Comment (corymhall) [inline docs/authentication-mode-migration.md:81]:
```suggestion
```ts
```

**Comment by flostadler** [comment]
The upgrade test failures are expected right now. They complain about a change in the aws provider, which does not skip the metadata API check by default now: https://github.com/pulumi/pulumi-aws/releases/tag/v6.37.1

Once this eks change is released we need to bump the baseline version and then they're green again

File: provider/provider_test.go [hunk 11]
```diff
@@ -142,61 +141,3 @@ func test(t *testing.T, dir string, opts ...providertest.Option) *providertest.P
 	)
 	return providertest.NewProviderTest(dir, opts...)
 }
-
```
Comment (flostadler) [inline provider/provider_test.go:145]: This retrieved the dependencies of the current version, but we need the deps of the baseline version

File: nodejs/eks/authenticationMode.test.ts [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+        expect(() => validateAuthenticationMode(args)).toThrowError(
+            "The 'accessEntries' property is not supported when 'authenticationMode' is not set.",
+        );
+    });
+
+    test.each([
+        [
+            {
+                authenticationMode: "CONFIG_MAP",
```
Comment (t0yv0) [inline nodejs/eks/authenticationMode.test.ts:109]: "CONFIG_MAP" is a literal not an enum val is this right?

File: examples/cluster-go/step1/main.go [hunk 13]
```diff
@@ -35,7 +35,7 @@ func main() {
 			return err
 		}
 
-		authMode := "API_AND_CONFIG_MAP"
+		authMode := eks.AuthenticationModeApiAndConfigMap
```
Comment (t0yv0) [inline examples/cluster-go/step1/main.go:38]: Oh nice.

File: examples/cluster-py/__main__.py [hunk 14]
```diff
@@ -53,7 +53,7 @@
                                 max_size=1,
                                 instance_type="t3.small"
                             ),
-                            authentication_mode=eks.AuthenticationMode.AP_I_AN_D_CONFI_G_MAP,
+                            authentication_mode=eks.AuthenticationMode.API_AND_CONFIG_MAP,
```
Comment (t0yv0) [inline examples/cluster-py/__main__.py:56]: Better.

File: nodejs/eks/cluster.ts [hunk 15]
```diff
@@ -1646,6 +1646,53 @@ export interface AccessPolicyAssociation {
...
+     * For IAM roles associated with self-managed Windows node groups. Allows the nodes to join the cluster.
+     */
+    EC2_WINDOWS: "EC2_WINDOWS",
+} as const;
+
+/**
+ * The authentication mode of the cluster. Valid values are `CONFIG_MAP`, `API` or `API_AND_CONFIG_MAP`
+ * See for more details:\nhttps://docs.aws.amazon.com/eks/latest/userguide/grant-k8s-access.html#set-cam
+ */
+export type AccessEntryType = (typeof AccessEntryType)[keyof typeof AccessEntryType];
```
Comment (t0yv0) [inline nodejs/eks/cluster.ts:1695]: Nice.

**Review by t0yv0 — APPROVED** [review]
Thank you!

File: examples/examples_nodejs_test.go [hunk 16]
```diff
@@ -685,3 +685,78 @@ func TestAccCNIAcrossUpdates(t *testing.T) {
 	t.Log("Ensuring that re-running `pulumi up` results in no changes and no spurious diffs")
 	pt.Up(optup.ExpectNoChanges())
 }
+
+func TestAccAuthenticationMode(t *testing.T) {
```
Comment (rquitales) [inline examples/examples_nodejs_test.go:689]: Note: these tests won't run in CI. You'd need to manually add them to the test matrix in our workflow files (eg. https://github.com/pulumi/pulumi-eks/blob/161e7af6803b58fea43ffdc3c6f9ee85cf658134/.github/workflows/run-acceptance-tests.yml#L422-L444).

**Review by rquitales — APPROVED** [review]
Changes looks good! Thanks for adding both unit and integration tests. Just a note that these tests won't currently run in CI. Approving to be non-blocking.

**Pull request merged**